uring/jvm/src/main/scala/fs2/io/uring/unsafe/UringSystem.scala

```diff
@@ -0,0 +1,250 @@
+      Resource.makeFull[IO, Int](poll =>
+        poll(exec(op, flags, rwFlags, fd, bufferAddress, length, offset)(release(_)))
+      )(release)
+
+    private def exec(
+        op: Byte,
+        flags: Int,
+        rwFlags: Int,
+        fd: Int,
+        bufferAddress: Long,
+        length: Int,
+        offset: Long
+    )(release: Int => IO[Unit]): IO[Int] = IO.cont {
+      new Cont[IO, Int, Int] {
+        def apply[F[_]](implicit
+            F: MonadCancelThrow[F]
+        ): (Either[Throwable, Int] => Unit, F[Int], IO ~> F) => F[Int] = { (resume, get, lift) =>
+          F.uncancelable { poll =>
+            println("[EXEC]: Entering exec")
+            println("THREAD:" + Thread.currentThread().getName)
+            val submit: IO[Short] = IO.async_[Short] { cb =>
+              register { ring =>
+                val id = ring.getSqe(resume)
+                val sq: UringSubmissionQueue = ring.getSq()
+                sq.enqueueSqe(op, flags, rwFlags, fd, bufferAddress, length, offset, id)
+                cb(Right(id))
+                println("[EXEC]: Leaving exec")
+              }
+            }
+
+            lift(submit)
+              .flatMap { id =>
+                F.onCancel(
+                  poll(get),
+                  lift(cancel(id)).ifM(
+                    F.unit,
+                    // if cannot cancel, fallback to get
+                    get.flatMap { rtn =>
+                      if (rtn < 0) F.raiseError(IOExceptionHelper(-rtn))
+                      else lift(release(rtn))
+                    }
+                  )
+                )
+              }
+              .flatTap(e => F.raiseWhen(e < 0)(IOExceptionHelper(-e)))
+          }
+
+        }
+      }
+    }
+
+    private[this] def cancel(
+        id: Short
+    ): IO[Boolean] =
+      IO.async_[Boolean] { cb =>
+        register { ring =>
+          val IORING_OP_ASYNC_CANCEL: Byte = 13
+
+          val wasCancel: Boolean =
+            !ring.getSq().enqueueSqe(IORING_OP_ASYNC_CANCEL, 0, 0, -1, 0, 0, 0, id)
+          cb(Right(wasCancel))
+        }
+      }
+  }
+
+  final class Poller private[UringSystem] (ring: UringRing) {
+
+    private[this] val sq: UringSubmissionQueue = ring.ioUringSubmissionQueue()
+    private[this] val cq: UringCompletionQueue = ring.ioUringCompletionQueue()
+
+    private[this] var pendingSubmissions: Boolean = false
+    private[this] val callbacks: Map[Short, Either[Throwable, Int] => Unit] =
+      Map.empty[Short, Either[Throwable, Int] => Unit]
+    private[this] val ids = BitSet(Short.MaxValue + 1)
+    private[this] var lastUsedId: Int = -1
+
+    private[this] def getUniqueId(): Short = {
+      val id = (lastUsedId + 1 until Short.MaxValue)
+        .find(!ids.contains(_))
+        .getOrElse(
+          (0 until lastUsedId)
+            .find(!ids.contains(_))
+            .getOrElse(
+              throw new RuntimeException("No available IDs")
+            )
+        )
+      lastUsedId = id
```

Comment:
If you use a Java `BitSet` instead of Scala `BitSet`, then you can use the `nextClearBit()` method to find the next available ID.
https://docs.oracle.com/javase/8/docs/api/java/util/BitSet.html#nextClearBit-int-